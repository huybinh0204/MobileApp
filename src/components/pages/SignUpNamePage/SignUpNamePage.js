import { yupResolver } from '@hookform/resolvers/yup';
import { useFocusEffect, useNavigation, useRoute } from '@react-navigation/native';
import React, { useCallback, useState } from 'react';
import { useForm } from 'react-hook-form';
import { LayoutAnimation } from 'react-native';
import { object, string } from 'yup';
import { Form, FormInput, FormSubmitButton } from '_components/atoms';
import { OnboardingScreenLayout } from '_components/organisms';
import { NAVIGATION, REGEX } from '_constants';
import strings from '_localization/index';
import { ButtonContainer, Container, Legend, Title } from './SignUpNamePage.styles';

const validationSchema = object().shape({
  firstName: string()
    .trim()
    .required(strings.signUp.requiredField)
    .matches(REGEX.name, strings.signUp.nameError)
    .min(1, strings.signUp.minError)
    .max(23, strings.signUp.maxError),
  lastName: string()
    .trim()
    .required(strings.signUp.requiredField)
    .matches(REGEX.name, strings.signUp.nameError)
    .min(2, strings.signUp.minError)
    .when('firstName', (firstName, schema) => {
      return schema.max(25 - firstName.length, strings.signUp.maxCombinedError);
    }),
});

const defaultValues = {
  firstName: '',
  lastName: '',
};

const SignUpNamePage = () => {
  const { params } = useRoute();
  const { navigate, addListener } = useNavigation();

  const [isContentVisible, setIsContentVisible] = useState(true);
  const formMethods = useForm({ defaultValues, resolver: yupResolver(validationSchema) });

  const firstNameValue = formMethods.watch('firstName');
  const lastNameValue = formMethods.watch('lastName');

  useFocusEffect(
    useCallback(() => {
      return addListener('beforeRemove', (event) => {
        event.preventDefault();
      });
    }, [addListener])
  );

  const showContent = () => {
    LayoutAnimation.easeInEaseOut();
    setIsContentVisible(true);
  };

  const hideContent = () => {
    LayoutAnimation.easeInEaseOut();
    setIsContentVisible(false);
  };

  const handleSubmit = ({ firstName, lastName }) => {
    navigate(NAVIGATION.auth.signUpBirthDate, { ...params, firstName, lastName });
  };

  return (
    <OnboardingScreenLayout step={6} showBackButton={false} testID="SignUpNamePage">
      <Container>
        {isContentVisible && (
          <>
            <Title>{strings.signUp.name.title}</Title>
            <Legend>{strings.signUp.name.legend}</Legend>
          </>
        )}
        <Form formMethods={formMethods}>
          <FormInput
            name="firstName"
            onFocus={hideContent}
            onSubmitEditing={showContent}
            placeholder={strings.signUp.name.firstNamePlaceholder}
            testID="firstName"
            textContentType="givenName"
          />
          <FormInput
            name="lastName"
            onFocus={hideContent}
            onSubmitEditing={showContent}
            placeholder={strings.signUp.name.lastNamePlaceholder}
            testID="lastName"
            textContentType="familyName"
          />
          <ButtonContainer>
            <FormSubmitButton
              disabled={firstNameValue === '' || lastNameValue === ''}
              onSubmit={handleSubmit}
              testID="signUpNameSubmitButton"
            >
              {strings.signUp.continue}
            </FormSubmitButton>
          </ButtonContainer>
        </Form>
      </Container>
    </OnboardingScreenLayout>
  );
};

export default SignUpNamePage;
