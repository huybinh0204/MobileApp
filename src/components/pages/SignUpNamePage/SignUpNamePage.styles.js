import styled from '@emotion/native';
import { BaseText } from '_components/atoms';

export const Container = styled.View(({ theme }) => ({
  flex: 1,
  padding: theme.spacing.m,
}));

export const Title = styled(BaseText)(({ theme }) => ({
  ...theme.typography.heading5,
  color: theme.colors.beta900,
  marginBottom: theme.spacing.xs,
}));

export const Legend = styled(BaseText)(({ theme }) => ({
  ...theme.typography.subContentRegular,
  color: theme.colors.beta500,
  marginBottom: theme.spacing.m,
}));

export const ButtonContainer = styled.View({
  marginTop: 'auto',
});
