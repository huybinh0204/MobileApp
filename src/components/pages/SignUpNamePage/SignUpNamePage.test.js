import { useNavigation } from '@react-navigation/native';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { NAVIGATION } from '_constants';
import strings from '_localization/index';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import { navigationMock } from '_test/mocks/navigation';
import SignUpNamePage from './SignUpNamePage';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}));

describe('SignUpNamePage', () => {
  const mockNavigation = navigationMock(jest.fn);

  beforeEach(() => {
    useNavigation.mockReturnValue(mockNavigation);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { toJSON } = render(
      <ComponentWithProviders component={SignUpNamePage} params={{ email: 'test@mail.com' }} />
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should trigger the change of screen if name and last name are corrects', async () => {
    const { getByText, getByTestId } = render(
      <ComponentWithProviders component={SignUpNamePage} params={{ email: 'test@mail.com' }} />
    );

    const firstNameInput = getByTestId('firstName');
    const lastNameInput = getByTestId('lastName');
    const submitButton = getByText(strings.signUp.continue);

    fireEvent.changeText(firstNameInput, 'First');
    fireEvent.changeText(lastNameInput, 'Last');

    expect(submitButton).toBeEnabled();

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    expect(mockNavigation.navigate).toBeCalledWith(NAVIGATION.auth.signUpBirthDate, {
      email: 'test@mail.com',
      firstName: 'First',
      lastName: 'Last',
    });
  });

  it('should not enable the continue button if some field is missing', async () => {
    const { getByText, getByTestId } = render(
      <ComponentWithProviders component={SignUpNamePage} params={{ email: 'test@mail.com' }} />
    );

    const firstNameInput = getByTestId('firstName');
    const submitButton = getByText(strings.signUp.continue);

    fireEvent.changeText(firstNameInput, 'First');

    expect(submitButton).toBeDisabled();

    fireEvent.press(submitButton);

    expect(mockNavigation.navigate).toBeCalledTimes(0);
  });

  it('should not navigate if first name + last name length is greatter than 25', async () => {
    const { findByText, getByText, getByTestId } = render(
      <ComponentWithProviders component={SignUpNamePage} params={{ email: 'test@mail.com' }} />
    );

    const firstNameInput = getByTestId('firstName');
    const lastNameInput = getByTestId('lastName');
    const submitButton = getByText(strings.signUp.continue);

    fireEvent.changeText(firstNameInput, 'ReallyLongFirstName');
    fireEvent.changeText(lastNameInput, 'LastName');
    fireEvent.press(submitButton);

    expect(await findByText(strings.signUp.maxCombinedError)).toBeTruthy();
    expect(mockNavigation.navigate).toBeCalledTimes(0);
  });
});
