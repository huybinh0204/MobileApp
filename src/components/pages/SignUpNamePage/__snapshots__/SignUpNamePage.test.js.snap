// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpNamePage should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTSafeAreaView
                        emulateUnlessSupported={true}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#FAFAFA",
                              "flex": 1,
                            },
                            undefined,
                          ]
                        }
                        testID="SignUpNamePage"
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "height": 108.5,
                                "justifyContent": "center",
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "width": "50%",
                                },
                                undefined,
                              ]
                            }
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "alignSelf": "center",
                                    "backgroundColor": "#F2F3F4",
                                    "borderColor": "#000000",
                                    "borderRadius": 10,
                                    "height": 9,
                                    "width": "100%",
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                collapsable={false}
                                style={
                                  Object {
                                    "backgroundColor": "#3D3397",
                                    "borderRadius": 10,
                                    "bottom": 0,
                                    "left": 0,
                                    "position": "absolute",
                                    "right": 0,
                                    "top": 0,
                                    "width": "0%",
                                  }
                                }
                              />
                            </View>
                          </View>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={
                            Array [
                              Array [
                                Object {
                                  "flex": 1,
                                },
                                undefined,
                              ],
                              Object {
                                "paddingBottom": 0,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "flex": 1,
                                  "padding": 58,
                                },
                                undefined,
                              ]
                            }
                          >
                            <Text
                              allowFontScaling={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#1F1D25",
                                    "fontFamily": "Poppins-SemiBold",
                                    "fontSize": 36,
                                    "lineHeight": 50.5,
                                    "marginBottom": 14.5,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              Whatâ€™s your name?
                            </Text>
                            <Text
                              allowFontScaling={false}
                              style={
                                Array [
                                  Object {
                                    "color": "#5F6A77",
                                    "fontFamily": "Poppins",
                                    "fontSize": 25.5,
                                    "lineHeight": 35.5,
                                    "marginBottom": 58,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              Your legal name is required to open an account.
                            </Text>
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginBottom": 87,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "borderRadius": 6,
                                      "borderWidth": 2,
                                      "flexDirection": "row",
                                      "width": "100%",
                                      "zIndex": 0,
                                    },
                                    Object {
                                      "borderColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <BVLinearGradient
                                  colors={
                                    Array [
                                      309817028,
                                      318208316,
                                    ]
                                  }
                                  endPoint={
                                    Object {
                                      "x": 0.75,
                                      "y": 0.5,
                                    }
                                  }
                                  locations={null}
                                  startPoint={
                                    Object {
                                      "x": 0.25,
                                      "y": 0.5,
                                    }
                                  }
                                  style={
                                    Array [
                                      Object {
                                        "borderRadius": 6,
                                        "height": "100%",
                                        "position": "absolute",
                                        "width": "100%",
                                        "zIndex": 1,
                                      },
                                      Object {
                                        "opacity": undefined,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  allowFontScaling={false}
                                  onLayout={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "bottom": 0,
                                        "fontFamily": "Poppins",
                                        "fontSize": 29,
                                        "left": 25.5,
                                        "lineHeight": 40,
                                        "position": "absolute",
                                        "zIndex": 2,
                                      },
                                      Object {
                                        "color": undefined,
                                        "transform": Array [
                                          Object {
                                            "translateY": undefined,
                                          },
                                          Object {
                                            "translateX": undefined,
                                          },
                                          Object {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  First name
                                </Text>
                                <TextInput
                                  onBlur={[Function]}
                                  onChangeText={[Function]}
                                  onFocus={[Function]}
                                  onSubmitEditing={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#1F1D25",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 29,
                                        "lineHeight": 40,
                                        "paddingBottom": 18,
                                        "paddingHorizontal": 25.5,
                                        "paddingTop": 43.5,
                                        "zIndex": 3,
                                      },
                                      undefined,
                                    ]
                                  }
                                  testID="firstName"
                                  textContentType="givenName"
                                  value=""
                                />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginBottom": 87,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "borderRadius": 6,
                                      "borderWidth": 2,
                                      "flexDirection": "row",
                                      "width": "100%",
                                      "zIndex": 0,
                                    },
                                    Object {
                                      "borderColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <BVLinearGradient
                                  colors={
                                    Array [
                                      309817028,
                                      318208316,
                                    ]
                                  }
                                  endPoint={
                                    Object {
                                      "x": 0.75,
                                      "y": 0.5,
                                    }
                                  }
                                  locations={null}
                                  startPoint={
                                    Object {
                                      "x": 0.25,
                                      "y": 0.5,
                                    }
                                  }
                                  style={
                                    Array [
                                      Object {
                                        "borderRadius": 6,
                                        "height": "100%",
                                        "position": "absolute",
                                        "width": "100%",
                                        "zIndex": 1,
                                      },
                                      Object {
                                        "opacity": undefined,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  allowFontScaling={false}
                                  onLayout={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "bottom": 0,
                                        "fontFamily": "Poppins",
                                        "fontSize": 29,
                                        "left": 25.5,
                                        "lineHeight": 40,
                                        "position": "absolute",
                                        "zIndex": 2,
                                      },
                                      Object {
                                        "color": undefined,
                                        "transform": Array [
                                          Object {
                                            "translateY": undefined,
                                          },
                                          Object {
                                            "translateX": undefined,
                                          },
                                          Object {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  Last name
                                </Text>
                                <TextInput
                                  onBlur={[Function]}
                                  onChangeText={[Function]}
                                  onFocus={[Function]}
                                  onSubmitEditing={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#1F1D25",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 29,
                                        "lineHeight": 40,
                                        "paddingBottom": 18,
                                        "paddingHorizontal": 25.5,
                                        "paddingTop": 43.5,
                                        "zIndex": 3,
                                      },
                                      undefined,
                                    ]
                                  }
                                  testID="lastName"
                                  textContentType="familyName"
                                  value=""
                                />
                              </View>
                            </View>
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginTop": "auto",
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                accessibilityState={
                                  Object {
                                    "disabled": true,
                                  }
                                }
                                accessible={true}
                                collapsable={false}
                                focusable={true}
                                onClick={[Function]}
                                onResponderGrant={[Function]}
                                onResponderMove={[Function]}
                                onResponderRelease={[Function]}
                                onResponderTerminate={[Function]}
                                onResponderTerminationRequest={[Function]}
                                onStartShouldSetResponder={[Function]}
                                style={
                                  Object {
                                    "alignSelf": "stretch",
                                    "backgroundColor": "#D4D7DB",
                                    "borderRadius": 10,
                                    "opacity": 1,
                                    "paddingHorizontal": 45.5,
                                    "paddingVertical": 31,
                                  }
                                }
                                testID="signUpNameSubmitButton"
                              >
                                <Text
                                  allowFontScaling={false}
                                  disabled={true}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#FFFFFF",
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 25.5,
                                        "lineHeight": 35.5,
                                        "textAlign": "center",
                                      },
                                      undefined,
                                    ]
                                  }
                                >
                                  Continue
                                </Text>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
